### DISCUSSION TITLE
Current state of WebRTC

### DISCUSSION THREAD

**Discussion by youngjoon-lee** [libp2p/rust-libp2p#3875] [Q&A] [answered]
Hi all,

I’m trying to test the connectivity of rust-libp2p for:
- standalone <> standalone
- standalone <> browser
- browser <> browser
After studying some materials from libp2p.io and after struggling with WebSocket transport (libp2p-websys-transport), I’m considering using WebRTC transport implemented in rust-libp2p. 

I would like to ask if the current state of WebRTC in rust-libp2p supports those connectivities. From [ROADMAP.md](https://github.com/libp2p/rust-libp2p/blob/master/ROADMAP.md#webrtc-support-browser-to-browser), it seems that only browser-to-server is supported currently, even if I use a relay server as described in https://docs.libp2p.io/concepts/transports/webrtc/. 

Also, I’d like to ask if I can start from these examples: 
- https://github.com/libp2p/rust-libp2p/tree/master/examples/relay-server
- https://github.com/libp2p/rust-libp2p/blob/master/transports/webrtc/examples/listen_ping.rs
It would be helpful if there is any example for Wasm with WebRTC. 

Thank you

**Comment by youngjoon-lee** [comment]
@thomaseizinger It seems that only WebRTC server-side has been completed by https://github.com/libp2p/rust-libp2p/pull/2622. As I understand, the WebRTC client-side is going to be implemented after resolving the WASM improvement roadmap. Did I understand it correctly?

https://github.com/libp2p/rust-libp2p/blob/master/ROADMAP.md#webrtc-support-browser-to-server
> We are currently implementing WebRTC for browser-to-server connectivity in https://github.com/libp2p/rust-libp2p/pull/2622. More specifically the server side. This will enable browser nodes to connect to rust-libp2p nodes where the latter only have self-signed TLS certificates. See https://github.com/libp2p/specs/pull/412 for in-depth motivation.
> 
> Long term we should enable rust-libp2p running in the browser via Wasm to use the browser's WebRTC stack. Though that should only happen after improved Wasm support, see below.

When I use js-libp2p for the client side, it worked well with the rust-libp2p server. If the WASM side of WebRTC is not completed yet as above, I’ll use js-libp2p for now. 

Thank you.

**Reply by youngjoon-lee** [reply, marked as answer]
I just posted a blog post about my rust-libp2p connectivity tests. 

https://oudwud.dev/posts/2305042345-exploring-rust-libp2p/

I’ll keep updating this post once I find something new. Please correct me if I’m wrong.

**Comment by p-shahi** [comment]
@youngjoon-lee So we've recently built an application that claims universal-connectivity i.e. showcases all the ways different libp2p implementations can connect with one another with their supported transports.
In this example, we connect rust-libp2p to a browser js-libp2p node using [webrtc-direct](https://github.com/libp2p/specs/blob/master/webrtc/webrtc-direct.md) and establish browser to server connectivity
We also establish QUIC connections between go-libp2p and rust-libp2p

The rust-libp2p node also acts as a Circuity Relay v2 Node and helps facilitate browser js-libp2p nodes to establish direct connections with other browser nodes using the [webrtc](https://github.com/libp2p/specs/blob/master/webrtc/webrtc.md) (private to private) transport

Here is the app: https://github.com/libp2p/universal-connectivity
We are currently in the progress of writing a tutorial for this

There's no Wasm involved in this example

**Reply by youngjoon-lee** [reply]
Awesome! Thank you. I’ll try to run it. 
It would be great if many users like me can find this example quickly from anywhere :)

**Reply by p-shahi** [reply]
Cool, please raise an issue in that repo if you have any trouble!

> It would be great if many users like me can find this example quickly from anywhere :)

I sympathize with you, will work to make sure it's easily discoverable

**Comment by thomaseizinger** [comment]
We will soon have a WebTransport implementation for WASM environments: https://github.com/libp2p/rust-libp2p/pull/4015

It should serve as a good blueprint in case somebody wants to attempt implementing WebRTC for WASM.

**Reply by youngjoon-lee** [reply]
👍 I have been keeping tracking of that issue. Happy to hear that it can be merged soon. 

> It should serve as a good blueprint in case somebody wants to attempt implementing WebRTC for WASM.

If this topic is in your backlog and if anyone in the team is not gonna take it, I can try to implement it.

**Reply by thomaseizinger** [reply]
> > It should serve as a good blueprint in case somebody wants to attempt implementing WebRTC for WASM.
> 
> If this topic is in your backlog and if anyone in the team is not gonna take it, I can try to implement it. 

I don't think we have an issue for it yet and I am not aware of anyone else working on it.

It would probably be good to create an issue to coordinate the work. It would be much appreciated! :)

**Reply by youngjoon-lee** [reply]
https://github.com/libp2p/rust-libp2p/issues/4111 I just opened an issue. I'll try to implement it if there is no plan by libp2p team.

**Comment by DougAnderson444** [comment]
Cross-referencing https://github.com/libp2p/rust-libp2p/pull/4248 as I'm working on the `web_sys` implementation for rust-libp2p WebRTC in the browser over there
